soccer/src/soccer/coach/coach_node.cpp

```diff
@@ -0,0 +1,140 @@
+        "/referee/play_state", 10,
+        [this](rj_msgs::msg::PlayState::SharedPtr msg) { play_state_callback(msg); });
+
+    world_state_sub_ = this->create_subscription<rj_msgs::msg::WorldState>(
+        "/vision_filter/world_state", 10,
+        [this](rj_msgs::msg::WorldState::SharedPtr msg) { world_state_callback(msg); });
+
+    // initialize all of the robot status subscriptions
+    // initialize our robots
+    for (int i = 0; i < kRobotsPerTeam; i++) {
+        robot_status_subs_[i] = this->create_subscription<rj_msgs::msg::RobotStatus>(
+            fmt::format("/radio/robot_status/robot_{}", i), 10,
+            [this](const rj_msgs::msg::RobotStatus::SharedPtr msg) {
+                ball_sense_callback(msg, true);
+            });
+    }
+
+    // initialize their robots
+    for (int i = kRobotsPerTeam; i < kRobotsPerTeam * 2; i++) {
+        robot_status_subs_[i] = this->create_subscription<rj_msgs::msg::RobotStatus>(
+            fmt::format("/radio/robot_status/robot_{}", i), 10,
+            [this](rj_msgs::msg::RobotStatus::SharedPtr msg) { ball_sense_callback(msg, false); });
+    }
+
+    current_play_state_.state = PlayState::State::Halt;
+    current_play_state_.restart = PlayState::Restart::Kickoff;
+    current_play_state_.our_restart = true;
+
+    rj_geometry_msgs::msg::Point temp_point;
+    temp_point.x = -1;
+    temp_point.y = -1;
+    current_play_state_.placement_point = temp_point;
+}
+
+void CoachNode::play_state_callback(rj_msgs::msg::PlayState::SharedPtr msg) {
+    current_play_state_ = *msg;
+    play_state_has_changed_ = true;
+}
+
+void CoachNode::world_state_callback(rj_msgs::msg::WorldState::SharedPtr msg) {
+    // EDGE-CASE NOTE: If robots from both teams are bordering the ball possession will likely
+    // switch repeatedly
+    if (!possessing_) {
+        for (rj_msgs::msg::RobotState robot_state : msg->our_robots) {
+            // There definitely has to be a better way, but this works...
+            if (rj_geometry::Point(robot_state.pose.position.x, robot_state.pose.position.y)
+                    .dist_to(rj_geometry::Point(msg->ball.position.x, msg->ball.position.y)) <
+                kRobotDiameter) {
+                possessing_ = true;
+                play_state_has_changed_ = true;
+                return;
+            }
+        }
+    } else {
+        for (rj_msgs::msg::RobotState robot_state : msg->their_robots) {
+            // See above...
+            if (rj_geometry::Point(robot_state.pose.position.x, robot_state.pose.position.y)
+                    .dist_to(rj_geometry::Point(msg->ball.position.x, msg->ball.position.y)) <
+                kRobotDiameter) {
+                possessing_ = false;
+                play_state_has_changed_ = true;
+                return;
+            }
+        }
+    }
+}
+
+void CoachNode::ball_sense_callback(rj_msgs::msg::RobotStatus::SharedPtr msg, bool our_team) {
```

Comment:
listen to this check warning
```suggestion
void CoachNode::ball_sense_callback(const rj_msgs::msg::RobotStatus::SharedPtr &msg, bool our_team) {
```
https://stackoverflow.com/questions/2627166/what-is-the-difference-between-a-const-reference-and-normal-parameter